### DISCUSSION TITLE
A way to organize files in "blog" folder?

### DISCUSSION THREAD

**Discussion by auracheeze** [facebook/docusaurus#8713] [General] [answered]
Is there a way to better organize large number of files in the _blog_ folder? At the moment I have about 100 files in the folder. I was wondering it will become really unwieldy when that number really becomes large.

Creating a subfolder inside _blog_ (e.g. _blog/archives_) does not work because it changes the slug (e.g. instead of `/blog/2008/05/15/post-1` it changes to `/blog/2008/05/15/archives/post-1`).

**Comment by slorber** [comment, marked as answer]
You can use the `slug` frontmatter explicitly to decouple the FS structure from the slug: this makes it possible to use the FS structure that you prefers.

We do this on Docusaurus blog: https://github.com/facebook/docusaurus/tree/main/website/blog
- first: used simple mdx files
- 2nd: used folders
- 3rd: create a `./releases` subfolder

![CleanShot 2023-03-02 at 12 10 41@2x](https://user-images.githubusercontent.com/749374/222412461-a12633a1-9d1b-47ea-9d41-89eb31813a57.png)

**Reply by auracheeze** [reply]
Thanks @slorber! I was hoping for an alternative that didn't involve updating every single article with the slug but I guess this is the best option.

**Reply by slorber** [reply]
I think we discussed the idea of a `createSlug` function (with @johnnyreilly ?) but I'm not too fan of the idea: it's good to be explicit and to ensure that your URLs do not change due to a code mistake, harming your site's SEO.

I think your URLs may not change if you decide to do: `/blog/2008/05/15/post-1.mdx` to `/blog/2008/05/15/post-1/index.mdx` because index is not part of the url segment? Didn't try but that could help. I also think you can replace `-` with `/`, we try to infer the post date from multiple filename patterns using both chars as possible separators.

**Reply by auracheeze** [reply]
The objective is to reduce the number of files/folders in the _blog_ directory. My idea was to group the articles in a year into one folder e.g.

- _/blog/2008-05-15-post-1.mdx_
- _/blog/2008-05-16-post-2.mdx_
- _/blog/2008-05-17-post-3.mdx_

to become

- _/blog/2008/2008-05-15-post-1.mdx_
- _/blog/2008/2008-05-16-post-2.mdx_
- _/blog/2008/2008-05-17-post-3.mdx_

so that there is only 1 folder _2008_ under _blog_.

At the same time, there shouldn't be too many sub-folders as well (e.g. _blog/2008/05/15_, _blog/2008/05/16_, _blog/2008/05/17_).

I think you are right about being explicit. It can be very easy to make a mistake. I've bitten the bullet and added a slug for every article. Appreciate your input 🙏

**Reply by slorber** [reply]
I think if you transform:

```
/blog/2008-05-15-post-1.mdx
/blog/2008-05-16-post-2.mdx
/blog/2008-05-17-post-3.mdx
```

to 

```
/blog/2008/05-15-post-1.mdx
/blog/2008/05-16-post-2.mdx
/blog/2008/05-17-post-3.mdx
```

or 

```
/blog/2008/05/15-post-1.mdx
/blog/2008/05/16-post-2.mdx
/blog/2008/05/17-post-3.mdx
```

Then the date can still be inferred and the URL would stay the same.

Didn't try but as far as I remember this should work

**Reply by slorber** [reply]
Regexp used:


```ts
const DATE_FILENAME_REGEX =
  /^(?<folder>.*)(?<date>\d{4}[-/]\d{1,2}[-/]\d{1,2})[-/]?(?<text>.*?)(?:\/index)?.mdx?$/;
```

Related unit tests:

```tsx
  it('parses folder name respecting date convention', () => {
    expect(
      parseBlogFileName('2021-05-12-announcing-docusaurus-two-beta/index.md'),
    ).toEqual({
      date: new Date('2021-05-12Z'),
      text: 'announcing-docusaurus-two-beta',
      slug: '/2021/05/12/announcing-docusaurus-two-beta',
    });
  });

  it('parses folder tree respecting date convention', () => {
    expect(
      parseBlogFileName('2021/05/12/announcing-docusaurus-two-beta/index.md'),
    ).toEqual({
      date: new Date('2021-05-12Z'),
      text: 'announcing-docusaurus-two-beta',
      slug: '/2021/05/12/announcing-docusaurus-two-beta',
    });
  });

  it('parses folder name/tree (mixed) respecting date convention', () => {
    expect(
      parseBlogFileName('2021/05-12-announcing-docusaurus-two-beta/index.md'),
    ).toEqual({
      date: new Date('2021-05-12Z'),
      text: 'announcing-docusaurus-two-beta',
      slug: '/2021/05/12/announcing-docusaurus-two-beta',
    });
  });

  it('parses nested folder tree respecting date convention', () => {
    expect(
      parseBlogFileName(
        '2021/05/12/announcing-docusaurus-two-beta/subfolder/file.md',
      ),
    ).toEqual({
      date: new Date('2021-05-12Z'),
      text: 'announcing-docusaurus-two-beta/subfolder/file',
      slug: '/2021/05/12/announcing-docusaurus-two-beta/subfolder/file',
    });
  });
```

**Reply by auracheeze** [reply]
My goodness! You are right! This deserves a mention in the [docs](https://docusaurus.io/docs/blog)! I'd like to make a PR to add this valuable piece of info.

**Reply by slorber** [reply]
@auracheeze please do :) 

We probably don't document this date filename convention properly but just expect people to find out it works fine out of the box

**Reply by auracheeze** [reply]
Upon closer look, I realised that the patterns are already listed in a summary section. I've added a small [PR](https://github.com/facebook/docusaurus/pull/8845) to draw more attention to this capability.
